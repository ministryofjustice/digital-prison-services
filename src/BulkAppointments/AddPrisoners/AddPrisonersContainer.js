import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import moment from 'moment'

import { H3 } from '@govuk-react/heading'
import Button from '@govuk-react/button'
import Page from '../../Components/Page'
import { Container, Divider } from './AddPrisonersContainer.styles'
import { ButtonContainer, ButtonCancel } from '../../Components/Buttons'

import {
  AppointmentTypeUsed,
  NumberOfAppointmentsEvent,
  RecordAbandonment,
  RecordRecurringAppointments,
  RecordWhenTheStartTimesHaveBeenAmended,
} from '../BulkAppointmentsGaEvents'

import { resetError, setLoaded, setAppointmentPrisoners, setBulkAppointmentsComplete } from '../../redux/actions'
import { appointmentType } from '../../types'
import routePaths from '../../routePaths'

import OffenderUpload from '../OffenderUpload/OffenderUpload'
import AddPrisoners from './AddPrisoners'
import AppointmentDetails from '../../AppointmentDetails/AppointmentDetails'
import ScrollToTop from '../../Components/ScrollToTop'
import { DATE_ONLY_FORMAT_SPEC } from '../../dateHelpers'

class AddPrisonerContainer extends Component {
  constructor(props) {
    super(props)
    this.state = { complete: false }
  }

  componentWillMount() {
    const { appointmentDetails, history } = this.props
    if (Object.keys(appointmentDetails).length === 0) {
      history.push(routePaths.bulkAppointments)
    }
  }

  componentDidMount() {
    const { setLoadedDispatch, resetErrorDispatch } = this.props
    setLoadedDispatch(true)
    resetErrorDispatch()
  }

  onBulkAppointmentsCreated({ appointments }) {
    const { dispatchBulkAppointmentsComplete, appointmentDetails } = this.props
    this.raiseEvents({ appointments })

    this.setState({
      complete: true,
      appointmentsCreated: appointments.length,
      lastAppointmentDetails: appointmentDetails,
    })

    dispatchBulkAppointmentsComplete()
  }

  onCancel(event) {
    const { raiseAnalyticsEvent, notmEndpointUrl } = this.props
    event.preventDefault()
    raiseAnalyticsEvent(RecordAbandonment())
    window.location = notmEndpointUrl
  }

  raiseEvents({ appointments }) {
    const { raiseAnalyticsEvent, agencyId, appointmentDetails } = this.props

    raiseAnalyticsEvent(NumberOfAppointmentsEvent({ total: appointments.length, agencyId }))
    raiseAnalyticsEvent(
      AppointmentTypeUsed({
        total: appointments.length,
        appointmentType: appointmentDetails.appointmentType,
      })
    )
    raiseAnalyticsEvent(
      RecordWhenTheStartTimesHaveBeenAmended({
        appointmentsDefaults: appointmentDetails,
        appointments,
      })
    )

    if (appointmentDetails.recurring) {
      const { repeats, times } = appointmentDetails
      RecordRecurringAppointments({
        repeatPeriod: repeats,
        count: times,
      })
    }
  }

  render() {
    const {
      appointmentDetails,
      now,
      resetErrorDispatch,
      handleError,
      history,
      dispatchAppointmentPrisoners,
      prisoners,
      agencyId,
    } = this.props

    const { complete, appointmentsCreated, lastAppointmentDetails } = this.state

    return (
      <Page title="Add prisoners" alwaysRender>
        {complete && (
          <ScrollToTop always>
            {lastAppointmentDetails.recurring ? (
              <H3> Appointments created </H3>
            ) : (
              <H3> {appointmentsCreated} appointments created </H3>
            )}
            <Divider />
            <AppointmentDetails {...lastAppointmentDetails} />
            <Divider />
            <ButtonContainer>
              <Button onClick={() => history.push(routePaths.bulkAppointments)}>Add another appointment</Button>
              <ButtonCancel onClick={e => this.onCancel(e)}>Back to home page</ButtonCancel>
            </ButtonContainer>
          </ScrollToTop>
        )}
        {!complete && (
          <React.Fragment>
            <p> Upload the prison numbers for all prisoners attending the below appointment. </p>
            <Container>
              <Divider />
              <AppointmentDetails {...appointmentDetails} />
              <Divider />
              <OffenderUpload
                showCancelButton={prisoners.length === 0}
                onCancel={event => this.onCancel(event)}
                onError={error => handleError(error)}
                agencyId={agencyId}
                onSuccess={data => {
                  resetErrorDispatch()
                  dispatchAppointmentPrisoners(data)
                }}
              />
              <Divider />
              <AddPrisoners
                onError={error => handleError(error)}
                resetErrors={resetErrorDispatch}
                onCancel={event => this.onCancel(event)}
                onSuccess={appointments => this.onBulkAppointmentsCreated({ appointments })}
                appointment={appointmentDetails}
                offenders={prisoners}
                now={now}
                date={appointmentDetails.date && appointmentDetails.date.format(DATE_ONLY_FORMAT_SPEC)}
                startTime={appointmentDetails.startTime}
                dispatchAppointmentPrisoners={dispatchAppointmentPrisoners}
              />
            </Container>
          </React.Fragment>
        )}
      </Page>
    )
  }
}
AddPrisonerContainer.propTypes = {
  resetErrorDispatch: PropTypes.func.isRequired,
  agencyId: PropTypes.string.isRequired,
  handleError: PropTypes.func.isRequired,
  raiseAnalyticsEvent: PropTypes.func.isRequired,
  notmEndpointUrl: PropTypes.string.isRequired,
  now: PropTypes.instanceOf(moment),
  appointmentDetails: appointmentType.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func.isRequired,
  }).isRequired,
  dispatchAppointmentPrisoners: PropTypes.func.isRequired,
  dispatchBulkAppointmentsComplete: PropTypes.func.isRequired,
  prisoners: PropTypes.arrayOf(PropTypes.object),
}

AddPrisonerContainer.defaultProps = {
  now: moment(),
  prisoners: [],
}

const mapStateToProps = state => ({
  agencyId: state.app.user.activeCaseLoadId,
  notmEndpointUrl: state.app.config.notmEndpointUrl,
  appointmentDetails: state.bulkAppointments.appointmentDetails,
  prisoners: state.bulkAppointments.prisoners,
})
const mapDispatchToProps = dispatch => ({
  resetErrorDispatch: () => dispatch(resetError()),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
  dispatchAppointmentPrisoners: prisoners => dispatch(setAppointmentPrisoners(prisoners)),
  dispatchBulkAppointmentsComplete: () => dispatch(setBulkAppointmentsComplete()),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddPrisonerContainer)
