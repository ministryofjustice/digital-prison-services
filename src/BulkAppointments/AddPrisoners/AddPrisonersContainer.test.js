import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'

import AddPrisonersContainer from './AddPrisonersContainer'

import {
  NumberOfAppointmentsEvent,
  AppointmentTypeUsed,
  RecordWhenTheStartTimesHaveBeenAmended,
} from '../BulkAppointmentsGaEvents'

const initialState = {
  app: {
    user: {
      activeCaseLoadId: 'LEI',
    },
    config: {
      notmEndpointUrl: '/',
    },
  },
  bulkAppointments: {
    appointmentDetails: {
      appointmentType: 'ACC',
      locationId: 1,
      appointmentTypeDescription: 'Activities',
      locationDescription: 'Gym',
      startTime: '2019-10-10T:21:00:00Z',
      comments: 'test comment',
      times: '1',
      recurring: true,
      repeats: 'DAILY',
      counts: '2',
    },
    offenders: [{ bookingId: 1 }, { bookingId: 2 }],
  },
}

describe('Add prisoners container', () => {
  const now = moment('2019-10-10T21:00:00Z')
  const store = {}
  const history = {}
  beforeEach(() => {
    store.getState = jest.fn()
    store.subscribe = jest.fn()
    store.dispatch = jest.fn()
    history.push = jest.fn()
    history.replace = jest.fn()
    store.getState.mockReturnValue(initialState)
  })
  const appointment = {
    appointmentType: 'ACC',
    locationId: 1,
    startTime: '2019-10-10T21:00:00Z',
    date: '2019-10-10',
  }

  it('should pass props correctly', () => {
    const wrapper = shallow(
      <AddPrisonersContainer
        now={now}
        store={store}
        handleError={() => {}}
        raiseAnalyticsEvent={() => {}}
        history={history}
      />
    )

    expect(wrapper.dive()).toMatchSnapshot()
  })

  it('should render completed view', () => {
    store.getState.mockReturnValue({
      ...initialState,
      bulkAppointments: {
        appointmentDetails: {
          ...initialState.bulkAppointments.appointmentDetails,
          recurring: false,
        },
      },
    })
    const wrapper = shallow(
      <AddPrisonersContainer
        now={now}
        store={store}
        handleError={() => {}}
        raiseAnalyticsEvent={() => {}}
        history={history}
      />
    )

    const component = wrapper.dive()

    component.instance().onBulkAppointmentsCreated({
      appointments: [
        { bookingId: 1, startTime: '2019-01-01T00:00:00Z' },
        { bookingId: 2, startTime: '2019-01-01T00:00:00Z' },
      ],
    })
    expect(component).toMatchSnapshot()
  })

  it('should render completed view for recurring', () => {
    const wrapper = shallow(
      <AddPrisonersContainer
        now={now}
        store={store}
        handleError={() => {}}
        raiseAnalyticsEvent={() => {}}
        history={history}
      />
    )

    const component = wrapper.dive()

    component.instance().onBulkAppointmentsCreated({
      appointments: [
        { bookingId: 1, startTime: '2019-01-01T00:00:00Z' },
        { bookingId: 2, startTime: '2019-01-01T00:00:00Z' },
      ],
    })
    expect(component).toMatchSnapshot()
  })

  it('should call raiseAnalyticsEvent with three different events', () => {
    const raiseAnalyticsEvent = jest.fn()
    const wrapper = shallow(
      <AddPrisonersContainer
        now={now}
        store={store}
        handleError={() => {}}
        raiseAnalyticsEvent={raiseAnalyticsEvent}
        history={history}
      />
    )

    const component = wrapper.dive()

    const appointments = [
      { bookingId: 1, startTime: '2019-01-01T00:00:00Z' },
      { bookingId: 2, startTime: '2019-01-01T00:00:00Z' },
    ]
    component.instance().onBulkAppointmentsCreated({
      appointments,
    })

    expect(raiseAnalyticsEvent.mock.calls[0][0]).toEqual(
      NumberOfAppointmentsEvent({ total: appointments.length, agencyId: 'LEI' })
    )

    expect(raiseAnalyticsEvent.mock.calls[1][0]).toEqual(
      AppointmentTypeUsed({
        total: appointments.length,
        appointmentType: appointment.appointmentType,
      })
    )

    expect(raiseAnalyticsEvent.mock.calls[2][0]).toEqual(
      RecordWhenTheStartTimesHaveBeenAmended({
        appointmentsDefaults: appointment,
        appointments,
      })
    )
  })
})
