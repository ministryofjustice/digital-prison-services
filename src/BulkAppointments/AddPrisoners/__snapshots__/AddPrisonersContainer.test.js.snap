// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add prisoners container should pass props correctly 1`] = `
<Connect(Page)
  alwaysRender={true}
  title="Add prisoners"
>
  <p>
     Upload the prison numbers for all prisoners attending the below appointment. 
  </p>
  <styled.div>
    <styled.div />
    <AppointmentDetails
      appointmentType="ACC"
      appointmentTypeDescription="Activities"
      comments="test comment"
      counts="2"
      endTime=""
      locationDescription="Gym"
      locationId={1}
      recurring={true}
      repeats="DAILY"
      startTime="2019-10-10T:21:00:00Z"
      times="1"
    />
    <styled.div />
    <OffenderUpload
      agencyId="LEI"
      onCancel={[Function]}
      onError={[Function]}
      onSuccess={[Function]}
      showCancelButton={true}
    />
    <styled.div />
    <AddPrisoners
      appointment={
        Object {
          "appointmentType": "ACC",
          "appointmentTypeDescription": "Activities",
          "comments": "test comment",
          "counts": "2",
          "locationDescription": "Gym",
          "locationId": 1,
          "recurring": true,
          "repeats": "DAILY",
          "startTime": "2019-10-10T:21:00:00Z",
          "times": "1",
        }
      }
      date={null}
      dispatchAppointmentPrisoners={[Function]}
      error=""
      now={"2019-10-10T21:00:00.000Z"}
      offenders={Array []}
      onCancel={[Function]}
      onError={[Function]}
      onSuccess={[Function]}
      resetErrors={[Function]}
      startTime="2019-10-10T:21:00:00Z"
    />
  </styled.div>
</Connect(Page)>
`;

exports[`Add prisoners container should render completed view 1`] = `
<Connect(Page)
  alwaysRender={true}
  title="Add prisoners"
>
  <withRouter(ScrollToTop)
    always={true}
  >
    <H3>
       
      2
       appointments created 
    </H3>
    <styled.div />
    <AppointmentDetails
      appointmentType="ACC"
      appointmentTypeDescription="Activities"
      comments="test comment"
      counts="2"
      endTime=""
      locationDescription="Gym"
      locationId={1}
      recurring={false}
      repeats="DAILY"
      startTime="2019-10-10T:21:00:00Z"
      times="1"
    />
    <styled.div />
    <styled.div>
      <ForwardRef
        disabled={false}
        onClick={[Function]}
        start={false}
      >
        Add another appointment
      </ForwardRef>
      <ForwardRef
        buttonColour="#dee0e2"
        buttonTextColour="#0b0c0c"
        disabled={false}
        onClick={[Function]}
        start={false}
      >
        Back to home page
      </ForwardRef>
    </styled.div>
  </withRouter(ScrollToTop)>
</Connect(Page)>
`;

exports[`Add prisoners container should render completed view for recurring 1`] = `
<Connect(Page)
  alwaysRender={true}
  title="Add prisoners"
>
  <withRouter(ScrollToTop)
    always={true}
  >
    <H3>
       Appointments created 
    </H3>
    <styled.div />
    <AppointmentDetails
      appointmentType="ACC"
      appointmentTypeDescription="Activities"
      comments="test comment"
      counts="2"
      endTime=""
      locationDescription="Gym"
      locationId={1}
      recurring={true}
      repeats="DAILY"
      startTime="2019-10-10T:21:00:00Z"
      times="1"
    />
    <styled.div />
    <styled.div>
      <ForwardRef
        disabled={false}
        onClick={[Function]}
        start={false}
      >
        Add another appointment
      </ForwardRef>
      <ForwardRef
        buttonColour="#dee0e2"
        buttonTextColour="#0b0c0c"
        disabled={false}
        onClick={[Function]}
        start={false}
      >
        Back to home page
      </ForwardRef>
    </styled.div>
  </withRouter(ScrollToTop)>
</Connect(Page)>
`;
